Allow constraints to be passed.

georgewwindsor opened the pull request:
specifically integrated maximum and minimum constraints for  frame-rate, width, and height.

MINIMUM resolution is 240x160 on my Iphone 5

Usage in JS code:
constraints = {video:
            {
              mandatory: {
                maxHeight: 160, // Provide your own width, height and frame rate here
                maxWidth: 240,
                maxFrameRate: 10
              },
              facingMode: CM.facingMode // 'environment'
            }
        ,audio: true};
getUserMedia(constraints, function(stream){
//low resolution stream returned.
});

m59peacemaker commented:
I'll try to test this today.

zxcpoiu commented:
anyone who are interested in video frame constraints, please take a look #267

georgewwindsor commented:
No problem, ill get those taken care of and get back to you. definately needs some cleanup. Also im not convinced framerate works. as far as android, is it simple a missing feature for max or it just hasnt been mapped correctly?

Max size is far more valuable to my team, because it reduces bandwidth, but i agree that both max and min being implemented would be ideal. ill take a look at passing some IOS min values.

georgewwindsor commented:
@zxcpoiu  I have cleaned up the code formatting, added 'minimum' constraint parsing, and changed the constraints to match the format that is specified in the documentation.

Please review and get back to me with any further changes you require to merge this in.

saghul commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,13 +268,98 @@ - (void)getUserVideo:(NSDictionary *)constraints
+         RTCMediaConstraints* betterConstraints = nil;
+         //If the video constraints exists, continue to parse its properties.  These properties match the format that is currently passed to chrome.
+
+         if(video != nil){
```
style: here and elsewhere: spacing: `if (condition) {`

saghul commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,13 +268,98 @@ - (void)getUserVideo:(NSDictionary *)constraints
+           NSDictionary *mandatoryParameters = video[@"mandatory"];
+             if(mandatoryParameters != nil){
+                 if([mandatoryParameters valueForKey:@"maxFrameRate"] != nil) {
+                     // The key existed...
```
pointless comment

saghul commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,13 +268,98 @@ - (void)getUserVideo:(NSDictionary *)constraints
+                     if ([[mandatoryParameters valueForKey:@"maxFrameRate"] isKindOfClass:[NSNumber class]]) {
+                         frameRateMax = [NSMutableString stringWithFormat:@"%@", (NSInteger)mandatoryParameters[@"maxFrameRate"]];
+                     }else{
+                         frameRateMax =mandatoryParameters[@"maxFrameRate"];
```
can't we just call this variable "maxFrameRate" ?

saghul commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,13 +268,98 @@ - (void)getUserVideo:(NSDictionary *)constraints
+               betterConstraints = [self defaultMediaStreamConstraints];
+             }
+         }else{
+         betterConstraints = [self defaultMediaStreamConstraints];
```
indentation is wrong here. Also, you can probably remove those else clauses by initializing constraints to the default and then overriding them if you find better ones.

georgewwindsor replied:
I dont think we should do this because the current constraints are just minimum, and not maximum, so you would have to nullify the opposite direction to make sure you weren't scoping your range too small, in some cases.  Ie you would have constraints that you didn't realize you had.

saghul commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,13 +268,98 @@ - (void)getUserVideo:(NSDictionary *)constraints
+         }
+         RTCAVFoundationVideoSource *videoSource = [self.peerConnectionFactory avFoundationVideoSourceWithConstraints:betterConstraints];
+
+      switch (videoDevice.position) {
```
broken indentation

georgewwindsor commented:
@zxcpoiu  @saghul  I have made the updates you requested, please review again and let me know if there any other changes needed.

georgewwindsor commented:
@zxcpoiu 
@saghul 

can you guys take a look at this? I have updated with the changes you requested...

saghul commented:
It looks pretty much the same as before to me. Many indentation and spacing issues. Also the 2 else clauses which can be simplified are still there.

georgewwindsor commented:
https://github.com/smiletime/react-native-webrtc/blob/master/ios/RCTWebRTC/WebRTCModule%2BRTCMediaStream.m

This is my updated code...looks pretty clean...? maybe i need to update the PR?

saghul commented:
> This is my updated code...looks pretty clean...? maybe i need to update the PR?

No, it doesn't look clean. Please review the diff instead of looking at the code in your editor.

georgewwindsor commented:
As far as the spacing issues, they are  a little off. as far the "if statements", they need to be there for logical reasons...if you guys would like to skype my ID is george.windsor5.

This requires discussion because im not convinced we are on the same page here. You CANNOT inherit the default "mandatory" constraints, because they MAY conflict with the constraints you pass in. simply "overwriting them" doesn't solve the problem, because it leaves constraints around that the "programmer" would not know were being applied.  

Please review more thoroughly, or explain what want more thoroughly.

zxcpoiu commented:
@georgewwindsor 

I understand your concern.
said, conflict may occurs when you pass only `max` constraints which are lower than the `default min` constraints, right?

it's true, but I think `to Pull Request a new feature` and `to change the default behavior` are two different things. It's better to make a PR's purpose as simple as possible, in this case, to add support for custom constraints.

because if you change the default behavior, it means all folks will have to change their settings accordingly.

besides, people who using `custom constraints` would probably knows constraints well, and are not likely to put constraints which will cause conflicts.

that's the reason I hope you can implement `min` and `max`, so that you can always put the right constraints to override the default one without any conflicts, and keep the default behavior the same.
everyone is happy.

looks like you did it, so to initialize the default constraints would not cause any problem as long as you understand the constraints and put the right one.

therefore, we can reduce lots of `if / else` logic to make the code `clean`.
`clean` generally is a subjective concept, the objective/relative speaking of clean, in this case, is `to make the code looks consistent to it's neighbor, to the entire repo, as possible`

zxcpoiu commented:
@georgewwindsor 

you aren't read closely what @saghul said, for those ` if(){ ` / `}else{`
should add space between them, like: `if () {` / `} else {`

zxcpoiu commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,12 +268,87 @@ - (void)getUserVideo:(NSDictionary *)constraints
+       RTCMediaConstraints* customConstraints = nil;
+       if (video != nil){
+         NSDictionary *mandatoryParameters = video[@"mandatory"];
+         if(mandatoryParameters != nil){
```
add space `if(` and `){`

zxcpoiu commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,12 +268,87 @@ - (void)getUserVideo:(NSDictionary *)constraints
+             NSMutableString *maxFrameRate = nil;
+             if ([[mandatoryParameters valueForKey:@"maxFrameRate"] isKindOfClass:[NSNumber class]]) {
+               maxFrameRate = [NSMutableString stringWithFormat:@"%@", (NSInteger)mandatoryParameters[@"maxFrameRate"]];
+             }else{
```
all `}else{` should be `} else {`

zxcpoiu commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,12 +268,87 @@ - (void)getUserVideo:(NSDictionary *)constraints
+           }
+           if ([mandatoryParameters valueForKey:@"maxHeight"] != nil) {
+             NSMutableString *maxHeight = nil;
+             if ([[mandatoryParameters valueForKey:@"maxHeight"] isKindOfClass:[NSNumber class]]) {
```
my opinion for type checking, is to do it on `js` side.
we can check or `toString`, do validation in javascript side
so on the native, we can just set constraint without validation hell.

it's ok to me if you didn't do it in js in the PR.

I can do it as a follow up. just to share my thoughts and my understanding to this repo's general preference.

reduce the line/logic/effort in native code not only make it easier to review/maintain, but also much more agile for adjust / refine / test / experiment. and reduce the size of necessary codes hard-codded in app.

it's the main reason we use `react-native`, isn't it?

saghul commented:
> As far as the spacing issues, they are a little off. as far the "if statements", they need to be there for logical reasons...if you guys would like to skype my ID is george.windsor5.

I don't understand what the problem is. Just make sure your code matches the style of the rest of the file.



> This requires discussion because im not convinced we are on the same page here. You CANNOT inherit the default "mandatory" constraints, because they MAY conflict with the constraints you pass in. simply "overwriting them" doesn't solve the problem, because it leaves constraints around that the "programmer" would not know were being applied.
> 
> Please review more thoroughly, or explain what want more thoroughly.

I'll rephrase: your code creates some custom constraints in `customConstraints`. All I said is that we should compact the code so you don't set it to the defaults in every "else" clause. Just reset it to null and check if there are no custom constraints at the end and set the defaults then. I never intended to imply merging defaults with those specified by the user.

georgewwindsor commented:
@saghul @zxcpoiu  I have made the changes requested. Please provide feedback.

1) Fixed Indentation
2) Removed type validation for strings/numbers etc. only strings allowed to be passed
3) created an mutable  mandatory Parameters with the default parameters.
4) No longer using defaultMediaStreamConstraints - Remove?

zxcpoiu commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,12 +268,43 @@ - (void)getUserVideo:(NSDictionary *)constraints
+    [mandatory setObject:@"720" forKey:@"maxHeight"];
+    [mandatory setObject:@"1280" forKey:@"maxWidth"];
+    [mandatory setObject:@"30" forKey:@"maxFrameRate"];
+    NSDictionary *video = constraints[@"video"];
```
we already have `id videoConstraints = constraints[@"video"];` above in https://github.com/smiletime/react-native-webrtc/blob/2836169d659e7f1fce66fce15ec9e386fae8f8ca/ios/RCTWebRTC/WebRTCModule%2BRTCMediaStream.m#L219

georgewwindsor replied:
Resolved.

zxcpoiu commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,12 +268,43 @@ - (void)getUserVideo:(NSDictionary *)constraints
+    NSMutableDictionary *mandatory = [[NSMutableDictionary alloc]init];
+    [mandatory setObject:@"720" forKey:@"maxHeight"];
+    [mandatory setObject:@"1280" forKey:@"maxWidth"];
+    [mandatory setObject:@"30" forKey:@"maxFrameRate"];
```
You can change `defaultMediaStreamConstraints`'s return type from `RTCMediaConstraints` to `NSDictionary` and initialize `mandatory` from it.

this might reviewed by @saghul

georgewwindsor replied:
agree, if you guys think i should use the defaultmediastreamconstraints. It dosnt really do anything tho.

georgewwindsor replied:
Resolved.

georgewwindsor commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,12 +268,43 @@ - (void)getUserVideo:(NSDictionary *)constraints
+    [mandatory setObject:@"1280" forKey:@"maxWidth"];
+    [mandatory setObject:@"30" forKey:@"maxFrameRate"];
+    NSDictionary *video = constraints[@"video"];
+    if (video != nil){
```
Resolved.

georgewwindsor commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,12 +268,43 @@ - (void)getUserVideo:(NSDictionary *)constraints
+    NSDictionary *video = constraints[@"video"];
+    if (video != nil){
+      NSDictionary *mandatoryParameters = video[@"mandatory"];
+      if(mandatoryParameters != nil){
```
Resolved.

georgewwindsor commented:
Please review the latest changes.

zxcpoiu commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,12 +265,39 @@ - (void)getUserVideo:(NSDictionary *)constraints
+          [mandatory setObject:minWidth forKey:@"minWidth"];
+        }
+      }
+    }
```
you did not use `defaultMediaStreamConstraints` at all.

if `mandatory` is not touched at the end, `mandatory` should fallback to `defaultMediaStreamConstraints`  for compatibility.

georgewwindsor commented:
updated. seems real solid.

zxcpoiu commented:
@saghul 

it seems ok to me.
please check is there anything I've missed

georgewwindsor commented:
there is a problem with the "id video = constraints" That is not a valid variable. I will be updating shortly.

      From: zxcpoiu <notifications@github.com>
 To: oney/react-native-webrtc <react-native-webrtc@noreply.github.com> 
Cc: georgewwindsor <georgewwindsor@yahoo.com>; Mention <mention@noreply.github.com>
 Sent: Wednesday, June 28, 2017 3:08 AM
 Subject: Re: [oney/react-native-webrtc] Allow constraints to be passed. (#288)
   
@saghulit seems ok to me.
please check is there anything I've missed—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

georgewwindsor commented:
I have added my fix, I think this is final, EXCEPT that you have to pass in strings. which does not match the android implementation. 
@saghul @zxcpoiu let me know if this works for you, I think we need the format to match on both android and ios though.


{ video:
            { mandatory: {
              minWidth: '500', // Provide your own width, height and frame rate here
              minHeight: '300',
              minFrameRate: '30'
            },

zxcpoiu commented:
How about use `toString` in JS side?
and accept only one type in native side?

I don't have video use case so can not be sure which is the best way.

georgewwindsor commented:
the native side only accepts a string right now. you can use a natural JS string ie var blah = '400' and its fine. toString would not be necessary. as long as people put in string values they will be fine.

I am indifferent to a solution. The only real problem is that the android version and the "documentation" for this project uses numbers and not strings. this will cause future confusion for people working on both...

      From: zxcpoiu <notifications@github.com>
 To: oney/react-native-webrtc <react-native-webrtc@noreply.github.com> 
Cc: georgewwindsor <georgewwindsor@yahoo.com>; Mention <mention@noreply.github.com>
 Sent: Wednesday, June 28, 2017 9:52 AM
 Subject: Re: [oney/react-native-webrtc] Allow constraints to be passed. (#288)
   
How about use `toString` in JS side?
and accept only one type in native side?

I don't have video use case so can not be sure which is the best way.
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

saghul commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,12 +265,43 @@ - (void)getUserVideo:(NSDictionary *)constraints
+        }
+      }
+    }
+    NSArray * mandatoryKeys = [mandatoryParams allKeys];
```
style: `*mandatoryKeys`

saghul commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,12 +265,43 @@ - (void)getUserVideo:(NSDictionary *)constraints
+      }
+    }
+    NSArray * mandatoryKeys = [mandatoryParams allKeys];
+    if([mandatoryKeys count]==0){
```
style: space after if, space between ==

saghul commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -38,16 +38,13 @@ @implementation WebRTCModule (RTCMediaStream)
+    NSMutableDictionary *mandatory = [[NSMutableDictionary alloc]init];
+    [mandatory setObject:@"720" forKey:@"maxHeight"];
+    [mandatory setObject:@"1280" forKey:@"maxWidth"];
+    [mandatory setObject:@"30" forKey:@"maxFrameRate"];
```
instead of repeating these string literals over and over I'd like to see a static constant.

saghul commented:
> EXCEPT that you have to pass in strings. which does not match the android implementation.

We need to have the implementations match, otherwise we cannot land it. Also, it must match the standard: https://www.w3.org/TR/mediacapture-streams/ So it's integers, NOT strings.

zxcpoiu commented:
@saghul is right.

* js spec: should pass a number ( type [long](https://heycam.github.io/webidl/#idl-long) )
* java: `videoCapturer.startCapture`, accept `int` as param.
* objc: `RTCMediaConstraints.initWithMandatoryConstraints` accept a `NSDictionary<NSString *, NSString *> *)`

I think the best way to match is to convert it on objc side. ios should only accept number then convert it to string. 

check
IF param is not `nil` AND `isKindOfClass:[NSNumber class]]` THEN
`[NSMutableString stringWithFormat:@"%d"`


this way, all the implementations are matched with the minimum changes. otherwise, we will have a verbose check on js.

georgewwindsor commented:
yup ill make those changes this evening.  i agree that its needed.

      From: zxcpoiu <notifications@github.com>
 To: oney/react-native-webrtc <react-native-webrtc@noreply.github.com> 
Cc: georgewwindsor <georgewwindsor@yahoo.com>; Mention <mention@noreply.github.com>
 Sent: Wednesday, June 28, 2017 10:39 PM
 Subject: Re: [oney/react-native-webrtc] Allow constraints to be passed. (#288)
   
@saghul is right.   
   - js spec: should pass a number ( type long )
   - java: videoCapturer.startCapture, accept int as param.
   - objc: RTCMediaConstraints.initWithMandatoryConstraints accept a NSDictionary<NSString *, NSString *> *)
I think the best way to match is to convert it on objc side. ios should only accept number then convert it to string.check
IF param is not nil AND isKindOfClass:[NSNumber class]] THEN
[NSMutableString stringWithFormat:@"%d"this way, all the implementations are matched with the minimum changes. otherwise, we will have a verbose check on js.—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

georgewwindsor commented:
Ok guys I have made the changes requested. I think we should be good to go.

georgewwindsor commented:
there are 2 other "constraint" related pull requests that can be removed, and about 10 "issues" related to constraints that can be removed.

zxcpoiu commented:
@georgewwindsor hey, you forgot the style spaces

zxcpoiu commented:
@georgewwindsor once this landed, will close related issues.

zxcpoiu commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -268,12 +262,43 @@ - (void)getUserVideo:(NSDictionary *)constraints
+      }
+    }
+    NSArray * mandatoryKeys = [mandatory allKeys];
+    if([mandatoryKeys count]==0){
```
spacing, should be `if ([mandatoryKeys count] == 0) {`

georgewwindsor replied:
got it.

zxcpoiu commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -266,14 +260,44 @@ - (void)getUserVideo:(NSDictionary *)constraints
+        }
+      }
+    }
+    NSArray * mandatoryKeys = [mandatory allKeys];
```
should be `*mandatoryKeys`

georgewwindsor replied:
done

zxcpoiu commented on ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:
```diff
@@ -293,7 +292,7 @@ - (void)getUserVideo:(NSDictionary *)constraints
     }
     NSArray * mandatoryKeys = [mandatory allKeys];
-    if([mandatoryKeys count]==0){
+    if([mandatoryKeys count] == 0){
```
should be `if ([mandatoryKeys count] == 0) {`

space between `if` and `(` 
space between `)` and `{`

georgewwindsor replied:
done. sorry my eyes are not trained to see it.:)

saghul commented:
@zxcpoiu This morning something clicked in my head. I was recently working on the Android permissions code, and noticed we do a similar thing as we do here.

I think we should move the parsing of resolutions to JS. That is, the native side would always get some 'mandatory' object with  'minHeight', 'minWidth' and 'minFps', which me make sure to set from JS. WDYT?

zxcpoiu commented:
@saghul 

Yes, that would be great.

ny preference is always put logic in js side and keep native side as simple as possible. just like we moved some default values from native to js.

putting logics and default values on js would speed up development process and easier for folks to test/customization/contributes.

so yes! as long as we keep `use case agnostic` and `flexible`  in mind.

we can use your PR as a follow up and close this PR and related issues

georgewwindsor commented:
kk let me know if you guys need me to make further changes.

georgewwindsor commented:
@zxcpoiu @saghul , if you guys dont have an existing PR for the way you would "like" to fix it, then i recommend merging this in. It is quite a blocking issue for literally dozens of peoples projects. It dosnt hurt me, im on a working fork.  i honestly think you guys need to prioritize getting this functionality merged in. This issue has been "open" for 18 months. Without it, your library is really only good for 2-way communication....and cannot support multiple parties at once with any reliability.

Additionally, I have made like 10 revisions doing exactly what you guys requested, so if you throw it away, I wont be contributing anything else, because you will have wasted my time.

zxcpoiu commented:
whoa, don't be mad mate.

we are only just a time limited enthusiast like you to dedicate our time in this repo.

it's unfair to said that "waste your time"
actually, why this PR rejected back and fourth several is mainly you did not look carefully and didn't modify what we repeatedly wrote in our review.

we all are not perfect and can not always toward the right direction. that's the reason we use github as a corporation platform to have discussions that make this repo better.

if you are also agree that doing constraints stuff in JS side is a better way, then you would not feel something "wasting time"

also, if you really agreed and fully understand why it's better to do it in JS, you can always fire a PR instead wrote emotional messages, isn't it?

saghul commented:
@georgewwindsor I won't enter any argument here. If this the attitude you want to have, good luck, I won't be looking at your PRs anymore. If you feel like calming down and finding the best way forward, let me know.

georgewwindsor commented:
ffs guys dont take my comments so personally. I stated some FACTS, there was a lot less emotion in it than you you guys are reading into it.  Why would i keep making PRs if you didnt merge them in? As I stated, I have a fork that I can keep working on, so none of this effects me. If you guys don't want help, thats fine. I have several other fixes for issues that I have developed, but ill just keep it on my fork.  Apparently you guys cant hear negativity without getting defensive.

zxcpoiu commented:
:expressionless: :expressionless: :expressionless:

we better stop meaningless discussion here. :cold_sweat:
I presume the sun will rise tomorrow and earth still circling. :alien:

saghul commented:
FTR, what needs to be done here is: perform all checks in JS land and make sure the "mandatory" section of the video constraints contains valid resolution / fps information, or the 720p/30fps defaults are applied.

The Android and iOS code would then read this, but without all the error checking.
